File: epi_judge_cpp/first_missing_positive_entry.cc
#include <vector>

#include "test_framework/generic_test.h"
using std::vector;

// A is passed by value argument, since we change it.
int FindFirstMissingPositive(vector<int> A) {
    size_t i = 0;
    while (i < A.size()) {
        if (A[i] > 0 && A[i] <= A.size() && A[A[i]-1] != A[i]) {
            std::swap(A[i], A[A[i]-1]);
        }
        else {
            i++;
        }
    }

    for (i = 0; i < A.size();i++) {
        if (A[i] != i + 1) 
            return i + 1;
    }

    return A.size()+1;
}

int main(int argc, char* argv[]) {
  std::vector<std::string> args{argv + 1, argv + argc};
  std::vector<std::string> param_names{"A"};
  return GenericTestMain(args, "first_missing_positive_entry.cc",
                         "first_missing_positive_entry.tsv",
                         &FindFirstMissingPositive, DefaultComparator{},
                         param_names);
}

File: epi_judge_python/first_missing_positive_entry.py
from typing import List

from test_framework import generic_test


def find_first_missing_positive(A: List[int]) -> int:
    n = len(A)
    for i in range(n):
        while 1 <= A[i] <= n and A[i] != A[A[i]-1]:
            b = A[i] - 1
            A[i], A[b] = A[b], A[i]

    for i in range(n):
        if A[i] != i + 1:
            return i + 1

    return n + 1


if __name__ == '__main__':
    exit(
        generic_test.generic_test_main('first_missing_positive_entry.py',
                                       'first_missing_positive_entry.tsv',
                                       find_first_missing_positive))
